### ISSUE TITLE
Rework of experimental options

### ISSUE THREAD

**Issue by ihhub** [ihhub/fheroes2#6161]
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.

### Describe the problem requiring a solution

Since there are always ongoing debates about whether to keep or remove certain experimental options we need to make a clear way how to deal with them.

### Describe the possible solution

I suggest to do these steps:
- [ ] add a configuration option to enable and disable all experimental options in the game. With the disabled state there is no way to change them or even load them with save files (controversial)
- [ ] save them in a separate file format which supports extension in the future
- [ ] handle issues with options being loaded from save files which update global state of these options within the game
- [ ] add a clear indicator that the game is being played with these options being enabled (an icon perhaps on the Adventure Map)
- [ ] propose a new dialog for these options to keep the same style of the game. Each option should follow the guide from [the wiki page](https://github.com/ihhub/fheroes2/wiki/Adding-a-new-option-into-the-game). We might stick to one set of sprites to turning on or off. Alternatively, we can use checkboxes but popup windows with a full description must be present. The page must contain a message telling that these options change the core gameplay and can lead to unexpected consequences.
- [ ] add an indicator that a save file contains enabled experimental options
- [ ] add an indicator that a high score was achieved with the enabled options
- [ ] add a question dialog for users which do not have these options enabled (global configuration flag) while they are loading a save file with the options
- [ ] add a way to show the options for campaign scenarios
- [ ] rename experimental options into something more game-style friendly

**Comment by ihhub** [comment]
Hi @zenseii , @Branikolog and @oleg-derevenetz , I jotted down my notes about experimental options rework. If you have even better ideas please feel free to explain.

**Comment by zenseii** [comment]
For reference here are all the current experimental options:
```
1. battle: show damage info
2. battle: deterministic events
3. battle: allow soft wait for troops

4. world: Scouting skill shows extended content info
5. world: allow to set guardian to objects
6. world: Windmills, Water Wheels and Magic Gardens can be captured

7. castle: allow guardians

8. heroes: allow buy a spellbook from Shrines
9. heroes: remember movement points when retreating or surrendering
10. heroes: allow to choose any primary skill in Arena

11. game: autosave will be made at the beginning of the day
12. game: offer to continue the game after victory condition
````

A bit off-topic, but I noticed these strings alongside the experimental ones when going through these in the PO file:
```
game: use evil interface
game: hide interface
```
Should they just be removed now that they are already part of the properly integrated options?

**Comment by zenseii** [comment]
Personally I think we should go for a dual approach: integrate some and move others to an external file.
The options are either modifications of the core rules of the game, or nice convenience features, and the majority are modifications.


My two arguments of having the least amount of options in the game is to not clutter up the game with options, and the other is that we will have less need to worry about finding the right icons or creating them.
The convenience options can all be either made default behavior or as options to them. They are:
```
battle: show damage info
game: autosave will be made at the beginning of the day
- This can be added as an extra autosave and rename the already existing one:
- "Autosave This Turn" and "Autosave Last Turn". The names are used as examples.
```
All the others can be made into external options that players that are curious enough can enable.

Two exceptions could be the deterministic battles and the continuing the game after victory. I believe we added the former ourselves and I believe many players would complain if we made it harder to enable? Still it seems like something you would either always have enabled or never enable so the small inconvenience of not being able to turn it on or off in-game should be insignificant. The same could be said about continuing the game after victory.

**Comment by oleg-derevenetz** [comment]
I believe that many of the experimental options can be either removed or integrated in the core.

> 1. battle: show damage info

Could be integrated and always turned on. IIRC OG has it too and it could be turned on/off via the Battle Options dialog (we have the Show Army Order) setting instead. BTW we can just always show the battle order and use this place in the Battle Options dialog to enable/disable this option, just like in the OG :)

> 2. battle: deterministic events

Can be integrated & always turned on.

> 3. battle: allow soft wait for troops

This option should remain "experimental" or removed, It might be useful in PvP hotseat, but AI currently has no idea how to use it. Although it would be a pity to remove it completely :)

> 7. castle: allow guardians

The same issue as with "soft wait". It's a pity to remove it.

> 9. heroes: remember movement points when retreating or surrendering

@LeHerosInconnu believes that the fact that the heroes in the original game get all the movement points back, even if they were hired back on the same day after they were defeated, is a bug and leads to the hit&run abuse. Maybe, but that's how the original game work. We can choose either logic, stick with it and remove this option.

> 4. world: Scouting skill shows extended content info
> 5. world: allow to set guardian to objects
> 6. world: Windmills, Water Wheels and Magic Gardens can be captured
> 8. heroes: allow buy a spellbook from Shrines
> 10. heroes: allow to choose any primary skill in Arena

Can be removed or remain "experimental".

> 11. game: autosave will be made at the beginning of the day
> 12. game: offer to continue the game after victory condition

Both can be removed.

As for the storage method for the experimental options, I believe that the current method isn't really so bad. We can just add some UI features, such as indication of experimental options presence and the ability to reset them all at once.

The real problem with experimental options is that it's hard to track them all during improvements/bugfixes. Since we have a very limited workforce, maybe really it worth to just remove arguable experimental options at all.

P.S. Although of course it would be a pity to remove at least two of them. I have made some efforts to properly support and fix bugs in some of them, including at least castle guardians, soft wait and extended scouting (by the way, the implementation of the latter looks somewhat overcomplicated).

**Comment by oleg-derevenetz** [comment]
I propose to make the "castle: allow guardians" option by default. After all, we already have changes in the core that affect the overall game balance even more - for example, the ability to hire non-upgraded units in the upgraded dwellings.

**Comment by zenseii** [comment]
I agree with most of @oleg-derevenetz's points. I do however wish to keep castle guardian as an option both personally and because the AI really needs to know how to use this, a fact which more than qualifies it as an "experimental" option. 

I also want to point out that we have users who prefer to use the autosave option. Would it be hard to make both worlds happy by having two auto saves? Heroes 4 has this exact system I described with two auto saves.

To summarize I think we can keep these three experimental options that @oleg-derevenetz mentioned:
> 3. battle: allow soft wait for troops
4. world: Scouting skill shows extended content info
7. castle: allow guardians

By reducing them to  just three options it should be easier to properly integrate them as options in the UI.

**Comment by zenseii** [comment]
Since we discussed moving the gameplay changing options away from the traditional game settings, maybe this could work for those specifically, if we decide to even keep them. I think it might start feeling a bit crowded in this menu at this point though.
<details>

![image](https://user-images.githubusercontent.com/12501091/201783893-4155e37d-c428-4922-8801-71db66479ca7.png)
</details>

**Comment by oleg-derevenetz** [comment]
Hi @zenseii the problem with these "experimental options" is that they are always at the roadside. They will always be "not good enough" because nobody uses them and nobody tests them, errors in them will accumulate as the logic develops. I believe that new features should either be integrated into the engine or removed.

**Comment by zenseii** [comment]
@oleg-derevenetz, this is a good point and I agree with this. It fits with the principle of the project to not bombard the player with options, but rather present them with some select polished features.

Hopefully we can get every onboard with this and then close all issues related to the specific experimental options dialog https://github.com/ihhub/fheroes2/issues?q=is%3Aissue+is%3Aopen+experimental+options
(there are also other issues not found there.)

**Comment by oleg-derevenetz** [comment]
@zenseii

> I also want to point out that we have users who prefer to use the autosave option. How much work would it be to create two auto saves? This would satisfy both worlds. Heroes 4 has this exact system I described with two auto saves and I find it convenient the times I use saves.

Is it really worth it to make 2 autosaves just so as not to wait for the end of the AI turn?

**Comment by zenseii** [comment]
> Is it really worth it to make 2 autosaves just so as not to wait for the end of the AI turn?

Sometimes you can have several sieges at many castles during the AI's turn, and having to redo these every time can be annoying. My other point was that some people like @LeHerosInconnu have previously noted that they use this feature, the same thing with the Arena +1 Knowledge choice. That's why it seemed best to find a way in the middle instead of either removing or integrating it.

You could always make the point that a player should just remember to save the game right after the AI player turn ends, and I could agree with that. I guess the same could also be said about the current autosave too though...

I also remember times when playing multiplayer in Heroes 4 when the connection broke and luckily we had the AutoSave from a later turn we could load up again and host from. Of course I'm sure fheroes2 network play will be a million times better and more stable, but it can never stop a power outage in your house xD

**Comment by oleg-derevenetz** [comment]
@zenseii What would you call these two autosave files? Preferably a shorter name.

**Comment by zenseii** [comment]
![image](https://user-images.githubusercontent.com/12501091/201789819-26e3d6f7-6961-43a6-88f8-7ea845cc94f4.png)

**Comment by oleg-derevenetz** [comment]
@zenseii The options, of course, are not particularly impressive :) Need to think a bit more. The advantage of names consisting solely from uppercase letters is that they are immediately stand out from the crowd.

**Comment by zenseii** [comment]
@oleg-derevenetz, yes the space is always a nice challenge.
How about these that keep capital letters but also distinguishes them. I prefer New/Old as they are more intuitive than 1st/2nd.
![image](https://user-images.githubusercontent.com/12501091/201792147-c85f50e6-8853-4bdc-8151-30d3c571c6a0.png)

**Comment by zenseii** [comment]
More alternatives:
![image](https://user-images.githubusercontent.com/12501091/201793050-a54941f5-b352-4473-8ba8-4176796ac731.png)
*NEW needs the margin to be smaller by a few more pixels.
![image](https://user-images.githubusercontent.com/12501091/201793256-57e4cea3-4ceb-40bf-860d-58cb48dbcaa7.png)

**Comment by oleg-derevenetz** [comment]
Hi @zenseii

> A bit off-topic, but I noticed these strings alongside the experimental ones when going through these in the PO file:

>> game: use evil interface
>> game: hide interface

> Should they just be removed now that they are already part of the properly integrated options?

These are not experimental options anymore, they are integrated in the GUI (see the Settings dialog that can be called from the adventure map screen).

**Comment by zenseii** [comment]
> These are not experimental options anymore, they are integrated in the GUI (see the Settings dialog that can be called from the adventure map screen).

Yes, this is what confused me when I found them as strings in the PO file:
![image](https://user-images.githubusercontent.com/12501091/201801632-3ea3605b-0df5-4822-8a44-45221653a47d.png)

I don't have the setup to quickly search for them in the project so I was wondering if anyone else could look at it and possibly remove the call to gettext `_()` for them.

**Comment by oleg-derevenetz** [comment]
OK, let's vote on the fate of the experimental options. My opinion is:

> battle: show damage info

Integrate instead the "show battle order" option (make the latter either always on, or controlled via config file only)

> battle: allow soft wait for troops

Remove.

> world: Scouting skill shows extended content info

Remove.

> world: allow to set guardian to objects

Don't know. I will abstain. 

> world: Windmills, Water Wheels and Magic Gardens can be captured

Abstain. 

> castle: allow guardians

Remove (see #6190). 

> heroes: allow buy a spellbook from Shrines

Abstain.

> heroes: remember movement points when retreating or surrendering

Abstain. 

> heroes: allow to choose any primary skill in Arena

Abstain. 

> game: autosave will be made at the beginning of the day

Integrate - always create a separate autosave files with different names at the end of the previous turn and at the beginning of a new turn as suggested by @zenseii. There will be a PR for that. 

> game: offer to continue the game after victory condition

Remove (see #6191).

@ihhub @LeHerosInconnu @Branikolog @zenseii your turn :)

**Comment by zenseii** [comment]
@oleg-derevenetz, do you suggest making a PR for the removal of every one of them? I just want to know if this will be the case for the scouting skill too which I believe has potential too.

Here are my votes:
 
> battle: show damage info

Integrate and make always default. It's less work to keep turn order as an option like it is right now.

> battle: allow soft wait for troops

Remove.

> world: Scouting skill shows extended content info

Remove.

> world: allow to set guardian to objects

Remove.

> world: Windmills, Water Wheels and Magic Gardens can be captured

Remove.

> castle: allow guardians

Remove.

> heroes: allow buy a spellbook from Shrines

Remove.

> heroes: remember movement points when retreating or surrendering

Integrate.

> heroes: allow to choose any primary skill in Arena

Remove. Such a feature can be added in a different new building.

> game: autosave will be made at the beginning of the day

Integrate - See  @oleg-derevenetz's comment about this.

> game: offer to continue the game after victory condition
 
Remove.

**Comment by oleg-derevenetz** [comment]
@zenseii yes, if option will be considered to be removed, I suggest to make a separate PR to remove each option, so they can be reverted/rewritten in future if there will be a need for that.

**Comment by Branikolog** [comment]
Hi, guys!
This thread becomes really hot! :) But at the same time I'm glad, that development becomes so involving and so many people are participating. :)

>battle: show damage info

Integrate. But instead of removing battle order (making it a default) I think it would be better to expand battle menu with new options. I know, lots of player hate battle order. Also, on 640x480 it covers creature sprites.

>battle: allow soft wait for troops

Remove. Definitely remove. :)

>world: Scouting skill shows extended content info

Remove.

>world: allow to set guardian to objects

Remove. We have nice spells for guarding mines.

>world: Windmills, Water Wheels and Magic Gardens can be captured

Remove. Resources have great influence on gameplay, so I'm against making them too affordable.

>castle: allow guardians

Remove.

>heroes: allow buy a spellbook from Shrines

Remove.

>heroes: remember movement points when retreating or surrendering

Integrate. The original mechanic (of recovering movement points after retreat) is heavily abused by players.

>heroes: allow to choose any primary skill in Arena

Remove. Originally this is an object rising 3 skills. We can implement in future new object for 4 skills.

>game: autosave will be made at the beginning of the day

Integrate. But this requires a lot of effort. Actually, I'm not satisfied with two ~same named files available in save list. I find it rather confusing. What if we simply show a separate window, while confirming AUTOSAVE, where a player can select between two options: `Previous` day or `The last day`? 

Otherwise, both savefiles could be visually separated from rest files, by another background texture or effect, another color or by placing them into a separate box within the window.

>game: offer to continue the game after victory condition

Remove. Useless option.

**Comment by oleg-derevenetz** [comment]
@Branikolog

> But this requires a lot of effort.

That's a problem. As all we know, there will be the first "real" release next month, and there is a lot of mandatory things to do. In fact, I even think to propose some sort of "feature freeze" at the end of the month to make sure that no new bugs will be introduced. Thus, the simpler solution is better.

**Comment by Branikolog** [comment]
Hi, @oleg-derevenetz 
 
> That's a problem. As all we know, there will be the first "real" release next month, and there is a lot of mandatory things to do. In fact, I even think to propose some sort of "feature freeze" at the end of the month to make sure that no new bugs will be introduced. Thus, the simpler solution is better.

Totally agreed! The option to load current day from the beginning is nice, but we can postpone its release a little bit.

**Comment by oleg-derevenetz** [comment]
@Branikolog so your proposal is to remove it for a while, but rework and add back later?

**Comment by Branikolog** [comment]
@oleg-derevenetz 

> @Branikolog so your proposal is to remove it for a while, but rework and add back later?

Yeah. Our upcoming release requires lots of other work. It would be more save, to introduce it without any untested features. The simpler - the better. After the release we can even keep players more interested, as we would release these "postponed" features, that have already been 80% ready, polished and discussed by us..

**Comment by zenseii** [comment]
The problem I see with removing it now is that we don't know when it will be properly added back unless we make it into a specific goal to have by x version. Looking at @oleg-derevenetz PR for implementing it, I don't think it's a problem moving on with that and if needed improving on this feature once if we think something could be better.

**Comment by Branikolog** [comment]
@zenseii 
I believe, in this particular case hardly anyone used this feature before, since it overwrites regular autosave, which people expect to happen at the end of the day. Introducing a feature with both savefiles available at the same menu would be some sort of innovation for our future players. I'd accept it even now, if we can promptly implement nice, well-looking design.

**Comment by zenseii** [comment]
> @zenseii I believe, in this particular case hardly anyone used this feature before, since it overwrites regular autosave, which people expect to happen at the end of the day. Introducing a feature with both savefiles available at the same menu would be some sort of innovation for our future players. I'd accept it even now, if we can promptly implement nice, well-looking design.

In that case I suggest removing it now and then taking a look at it at a later point, since we don't have the time to make the perfect solution now.

**Comment by ihhub** [comment]
I will give my opinion and replies tomorrow as I am very busy at the moment :(

**Comment by LeHerosInconnu** [comment]
Hello everyone,

- heroes: remember movement points when retreating or surrendering
Integrate.
For me, a bug in the original game.

- battle: show damage info
Integrate.
Can be improved.

- world: Windmills, Water Wheels and Magic Gardens can be captured
Integrate.
A must have as it removes boring and time consuming micromanagement from the game.
I want my heroes to do strategic actions not maintenance.
Perhaps as an option to activate for those nostalgic of the way the original game works.

- world: Scouting skill shows extended content info
Remove this version.
Can be done better.

- heroes: allow buy a spellbook from Shrines
Integrate.
And the cost of the magic book should be the same as in the Mage Guild (500 gold) for all types of shrines.

- heroes: allow to choose any primary skill in Arena
Integrate.
Spell Power is useless if the hero does not have enough spell points to use it effectively.

- game: autosave will be made at the beginning of the day
Integrate, with both files (https://github.com/ihhub/fheroes2/pull/6192).
I suggest to keep the name "AUTOSAVE" in uppercase for the file corresponding to the same autosave file of the original game and to name the file "AUTOTURN" in uppercase for the file saved after the actions of the other players.
The name "AUTOTURN" refers to the save made automatically at the beginning of the player's turn.

- game: offer to continue the game after victory condition
Integrate.
I use this option for testing, for example to see what constructions have been built, what spells the Mage Guild has, what creatures are available for recruitment in the last opponent's city, and also to check the obelisks and the Ultimate Artifact.
Some players, notably @1GodRage, also like to have the possibility to continue the scenario after the victory, let them have this pleasure.
In my opinion, this should remain an option to be activated so as not to "pollute" the game experience of those who are not interested in it.

- world: allow to set guardian to objects
Remove.
For me it's a waste of creatures because they don't benefit from the bonuses given by a hero during a combat.
Also, it can be a bit abused by the players (hero movement, troop transfer), adding, therefore, an additional micromanagement.

- castle: allow guardians
Remove.
Can be done better.

- battle: allow soft wait for troops
Remove.

**Comment by Mr-Bajs** [comment]
Even though i like several of these options as a player.

Why not just remove most of experimental features from the 1.0 and postpone the experimental features and bundle them to the expansion Resurrection 1.3 release, seams to to more inline with the goal of 1.0?

https://github.com/ihhub/fheroes2/wiki/Project-roadmap

After 1.0 maybe the should be just be 1.0 settings that is more less just original Homm2 game rules. And a bundled experimental setting for an future expansion, i suppose it would be easier to get an more coherent experience during development if everyone plays the game with the same settings. 

You either play with fheroes 1.0 ruleset or an post 1.0 release added Resurrection experimental setting.

**Comment by oleg-derevenetz** [comment]
OK, let's summarize the interim results in this table:

| - | oleg-derevenetz | zenseii |Branikolog | LeHerosInconnu | ihhub |
| --- | --- | --- | --- | --- | --- |
| battle: show damage info | + | + | + | + | N/A |
| battle: allow soft wait for troops | - | - | - | - | N/A |
| world: Scouting skill shows extended content info | - | - | - | - | N/A |
| world: allow to set guardian to objects | ? | - | - | - | N/A |
| world: Windmills, Water Wheels and Magic Gardens can be captured | ? | - | - | + | N/A |
| castle: allow guardians | - | - | - | - | N/A |
| heroes: allow buy a spellbook from Shrines | ? | - | - | + | N/A |
| heroes: remember movement points when retreating or surrendering | ? | + | + | + | N/A |
| heroes: allow to choose any primary skill in Arena | ? | - | - | + | N/A |
| game: autosave will be made at the beginning of the day | + | - | - | + | N/A |
| game: offer to continue the game after victory condition | - | - | - | + | N/A |

The legend is:

* `+` - integrate in the release as a part of the game core;
* `-` - remove from the release (this includes the option "remove for a while, but then rework and add back");
* `?` - abstain;
* `N/A` - haven't voted yet.

So it seems that we have two favorites so far ("battle: show damage info" and "heroes: remember movement points when retreating or surrendering") and two outsiders ("battle: allow soft wait for troops" and "castle: allow guardians"). The results for the rest are contradictory, and therefore most likely they will be removed from the 1.0 (but, perhaps, they will be reworked and added later, but not as "experimental options" anymore, I believe we've had enough of this).

**Comment by oleg-derevenetz** [comment]
@Branikolog

>> battle: allow soft wait for troops
>
> Remove. Definitely remove. :)

BTW, why you don't like this option so much? The mechanics of waiting dramatically increases the flexibility of using troops, especially slow ones. It was an undoubted innovation in HoMM3.

**Comment by Alucard648** [comment]
battle: show damage info - Integrate - A must have to avoid heavy hassle for damage calculation.
battle: allow soft wait for troops - Postpone for expansion - requires AI adaptation.
world: Scouting skill shows extended content info - Postpone for expansion - Again AI should handle option well. Also heavy buff already OP warlocks that can how quickly gain Expert Scouting for exact neutral numbers and vacuum map rapidly. Hello Dragons on week 4!
world: allow to set guardian to objects - Remove - Results in abuse, involving spending a night in mine for logistics boost.
world: Windmills, Water Wheels and Magic Gardens can be captured - Keep - I hate migromanadgement.
castle: allow guardians - Postpone for expansion - Again, heroes 3-like feature better to be in expansion.
heroes: allow buy a spellbook from Shrines - Keep - Nothing more frustrating than forgetting to buy spellbook in castle mage guild ...
heroes: remember movement points when retreating or surrendering - Remove - Definitely an abuse that can be devastating. Think Alteris 1&2. Heroes should have 0 MP on buyout from retreat.
heroes: allow to choose any primary skill in Arena - Keep - A must for heroes that have low chance on getting KNW on level up
game: autosave will be made at the beginning of the day - keep - Autosaves should be done on both ending turn and starting next day.
game: offer to continue the game after victory condition - Remove - Useless.

**Comment by ihhub** [comment]
Hi everyone!

This is my vision what we should do with the remaining options:
> 1. battle: show damage info
Integrate into the game with an icon within battle options. I will take care of it.

> 2. battle: allow soft wait for troops
Remove from 1.0 release. We can put it back only after teaching AI to use this option and if we won't add monsters with an ability to wait.

> 3. world: Scouting skill shows extended content info
Remove. Useless option and a combination of skills and artifacts give the same effect.

> 4. world: allow to set guardian to objects
Remove. As pointed out we have spells for this. We can instead add new spells by doing it which will be integration into the game. This options makes heroes 2 more like heroes 4.

> 5. world: Windmills, Water Wheels and Magic Gardens can be captured
Remove. We can reintroduce it only if we add visual effects / new buildings to highlight such objects on the map. From user perspective if you see an object A you expect behavior X, not behavior X or Y.

> 6. castle: allow guardians
The same as above. We need to add visual hints for this option.

> 7. heroes: allow buy a spellbook from Shrines
The same as above. We need to add visual hints for this option.

> 8. heroes: remember movement points when retreating or surrendering
Integrate into the game.

> 9. heroes: allow to choose any primary skill in Arena
I am okay to remove or make it permanent.

> 10. game: autosave will be made at the beginning of the day
Remove for now. Having 2 save files is not a user-friendly experience. Heroes 4 have such feature.

> 11. game: offer to continue the game after victory condition
Remove. Games are designed to be completed when you archive your goal. At least for most of cases. For HoMM after the completion there not much to do.

As a conclusion: we make `battle: show damage info` and `heroes: remember movement points when retreating or surrendering` a part of the game. The rest of options I suggest to remove either completely or at least the code related to UI.

**Comment by ihhub** [comment]
Hi @oleg-derevenetz , I believe we left (based on majority of votes) with this set of actions:
- integrate `battle: show damage info` option which I am going to take care
- remove the rest of options except `game: autosave will be made at the beginning of the day` which we have 50/50 votes
